crates/ra_ide/src/hover.rs

```diff
@@ -335,6 +346,277 @@ fn hover_for_definition(db: &RootDatabase, def: Definition) -> Option<Markup> {
     }
 }
 
+// Rewrites a markdown document, resolving links using `callback` and additionally striping prefixes/suffixes on link titles.
+fn map_links<'e>(
+    events: impl Iterator<Item = Event<'e>>,
+    callback: impl Fn(&str, &str) -> (String, String),
+) -> impl Iterator<Item = Event<'e>> {
+    let mut in_link = false;
+    let mut link_target: Option<CowStr> = None;
+
+    events.map(move |evt| match evt {
+        Event::Start(Tag::Link(_link_type, ref target, _)) => {
+            in_link = true;
+            link_target = Some(target.clone());
+            evt
+        }
+        Event::End(Tag::Link(link_type, _target, _)) => {
+            in_link = false;
+            Event::End(Tag::Link(link_type, link_target.take().unwrap(), CowStr::Borrowed("")))
+        }
+        Event::Text(s) if in_link => {
+            let (link_target_s, link_name) = callback(&link_target.take().unwrap(), &s);
+            link_target = Some(CowStr::Boxed(link_target_s.into()));
+            Event::Text(CowStr::Boxed(link_name.into()))
+        }
+        Event::Code(s) if in_link => {
+            let (link_target_s, link_name) = callback(&link_target.take().unwrap(), &s);
+            link_target = Some(CowStr::Boxed(link_target_s.into()));
+            Event::Code(CowStr::Boxed(link_name.into()))
+        }
+        _ => evt,
+    })
+}
+
+/// Rewrite documentation links in markdown to point to an online host (e.g. docs.rs)
+fn rewrite_links(db: &RootDatabase, markdown: &str, definition: &Definition) -> String {
```

Comment:
Yup, I think this is the right interface!

Long-term, we should implement some kind of DOM model for the documentation, so that the client decides how to best render it, and can choose, for example, between doc.rs links or links to the source code. But, as a first step, just hard-coding docs.rs seems like a good choice!

Reply:
Just to be clear, it's not hardcoded to docs.rs, as long as `html_root_url` is specified  this will correctly link to wherever the docs are hosted.

I agree, I initially started supporting the ability to provide `file://` URLs to local documentation, but AFAIK VSCode will always open those links internally, and that wasn't behaviour I saw value in.